planning/mission_planner/src/mission_planner/mission_planner.cpp

```diff
@@ -299,25 +367,171 @@ void MissionPlanner::on_set_mrm_route(
+      ResponseCode::ERROR_PLANNER_UNREADY, "Normal route is not set.");
+  }
+  if (!mrm_route_) {
+    change_state(RouteState::Message::SET);
+    throw component_interface_utils::ServiceException(
+      ResponseCode::ERROR_PLANNER_UNREADY, "Mrm route is not set.");
+  }
+
+  // check route safety
+  if (checkRerouteSafety(*mrm_route_, *normal_route_)) {
+    clear_mrm_route();
+    change_route(*normal_route_);
+    change_state(RouteState::Message::SET);
+    res->success = true;
+    is_emergency_ = false;
+    return;
+  }
+
+  // reroute with normal goal
+  const auto new_route =
+    create_route(odometry_->header, normal_route_->goal_pose, normal_route_->allow_modification);
+
+  if (new_route.segments.empty()) {
+    change_route(*normal_route_);
+    change_state(RouteState::Message::SET);
+    throw component_interface_utils::ServiceException(
+      ResponseCode::ERROR_PLANNER_FAILED, "The planned route is empty.");
+  }
+
+  // check new route safety
+  if (new_route.segments.empty() || !checkRerouteSafety(*mrm_route_, new_route)) {
+    // failed to create a new route
+    RCLCPP_ERROR_THROTTLE(get_logger(), *get_clock(), 5000, "Reroute with normal goal failed.");
+    change_mrm_route(*mrm_route_);
+    change_route(*normal_route_);
+    res->success = false;
+    is_emergency_ = true;
+  } else {
+    clear_mrm_route();
+    change_route(new_route);
+    res->success = true;
+    is_emergency_ = false;
+  }
+
+  change_state(RouteState::Message::SET);
 }
 
 void MissionPlanner::on_modified_goal(const ModifiedGoal::Message::ConstSharedPtr msg)
 {
-  // TODO(Yutaka Shimizu): reroute if the goal is outside the lane.
-  arrival_checker_.modify_goal(*msg);
+  using ResponseCode = autoware_adapi_v1_msgs::srv::SetRoute::Response;
+
+  if (state_.state != RouteState::Message::SET) {
+    throw component_interface_utils::ServiceException(
+      ResponseCode::ERROR_INVALID_STATE, "The route hasn't set yet. Cannot reroute.");
+  }
+  if (!odometry_) {
+    throw component_interface_utils::ServiceException(
+      ResponseCode::ERROR_PLANNER_UNREADY, "The vehicle pose is not received.");
+  }
+  if (!normal_route_) {
+    throw component_interface_utils::ServiceException(
+      ResponseCode::ERROR_PLANNER_UNREADY, "Normal route is not set.");
+  }
+
+  // set to changing state
+  change_state(RouteState::Message::CHANGING);
+
+  if (normal_route_->uuid == msg->uuid) {
+    const auto new_route = create_route(msg->header, msg->pose, normal_route_->allow_modification);
+    if (new_route.segments.empty()) {
+      change_route(*normal_route_);
+      change_state(RouteState::Message::SET);
+      throw component_interface_utils::ServiceException(
+        ResponseCode::ERROR_PLANNER_FAILED, "The planned route is empty.");
+    }
+    change_route(new_route);
+    change_state(RouteState::Message::SET);
+    arrival_checker_.modify_goal(*msg);
+    return;
+  }
+
+  if (mrm_route_ && mrm_route_->uuid == msg->uuid) {
```

Comment:
```suggestion
  if (is_emergency_ && mrm_route_ && mrm_route_->uuid == msg->uuid) {
```
?

Reply:
maybe it is not needed if alwawy set mrm_route_ when emegency and clear mrm_route when not emegeryc.